### PR TITLE
Inbound fixes

### PR DESCRIPTION
**Pull request by hdevalence**
Fixes some oversights from #1071.

### CODE REVIEW

File: zebra-network/src/peer_set/candidate_set.rs [hunk 1]
```diff
@@ -106,7 +106,7 @@ where
         // existing peers, but we don't make too many because update may be
         // called while the peer set is already loaded.
         let mut responses = FuturesUnordered::new();
-        for _ in 0..2usize {
+        for _ in 0..1usize {
```
Comment (dconnolly) [inline zebra-network/src/peer_set/candidate_set.rs:110]: Is this just length 1?

File: zebrad/src/components/inbound.rs [hunk 2]
```diff
@@ -73,61 +73,62 @@ impl Service<zn::Request> for Inbound {
...
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        use oneshot::error::TryRecvError;
-        match self.network_setup.take() {
-            // If we're waiting for setup, check if we became ready
-            Some(mut rx) => match rx.try_recv() {
+        // Check whether the network setup is finished, but don't wait for it to
+        // become ready before reporting readiness. We expect to get it "soon",
+        // and reporting unreadiness might cause unwanted load-shedding, since
+        // the load-shed middleware is unable to distinguish being unready due
+        // to load from being unready while waiting on setup.
```
Comment (dconnolly) [inline zebrad/src/components/inbound.rs:80]: Bummer

File: zebrad/src/components/inbound.rs [hunk 3]
```diff
@@ -73,61 +73,62 @@ impl Service<zn::Request> for Inbound {
...
-                    self.state.poll_ready(cx),
-                    self.outbound.as_mut().unwrap().poll_ready(cx),
-                ) {
-                    (Poll::Ready(Err(e)), _) | (_, Poll::Ready(Err(e))) => Poll::Ready(Err(e)),
-                    (Poll::Pending, _) | (_, Poll::Pending) => Poll::Pending,
-                    (Poll::Ready(Ok(())), Poll::Ready(Ok(()))) => Poll::Ready(Ok(())),
+                Err(e @ TryRecvError::Closed) => {
+                    // In this case, report that the service failed, and put the
+                    // failed oneshot back so we'll fail again in case
+                    // poll_ready is called after failure.
```
Comment (dconnolly) [inline zebrad/src/components/inbound.rs:95]: 👍

File: zebrad/src/components/inbound.rs [hunk 4]
```diff
@@ -73,61 +73,62 @@ impl Service<zn::Request> for Inbound {
...
-                const MAX_ADDR: usize = 1000; // bitcoin protocol constant
-                peers.truncate(MAX_ADDR);
-                async { Ok(zn::Response::Peers(peers)) }.boxed()
-            }
+            zn::Request::Peers => match self.address_book.as_ref() {
+                Some(addrs) => {
+                    // We could truncate the list to try to not reveal our entire
+                    // peer set. But because we don't monitor repeated requests,
+                    // this wouldn't actually achieve anything, because a crawler
+                    // could just repeatedly query it.
```
Comment (dconnolly) [inline zebrad/src/components/inbound.rs:124]: 👍

File: zebra-network/src/peer_set/candidate_set.rs (hunk 1 above)
Comment (hdevalence) [reply zebra-network/src/peer_set/candidate_set.rs:110]: Yes, we could also cut out the loop entirely and just make one request, at the cost of it being more work to add fanout back in later.

File: zebrad/src/components/inbound.rs (hunk 2 above)
Comment (hdevalence) [reply zebrad/src/components/inbound.rs:80]: On the other hand, doing this means that we can start responding to state queries immediately.

File: zebra-network/src/peer_set/candidate_set.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-network/src/peer_set/candidate_set.rs:110]:
```suggestion
        // Yes this loops only once (for now), until we add fanout back.
        for _ in 0..1usize {
```

Comment (hdevalence) [reply zebra-network/src/peer_set/candidate_set.rs:110]: Happy to do this or just cut out the loop, whichever is preferable.

File: zebrad/src/components/inbound.rs [hunk 5]
```diff
@@ -73,61 +73,62 @@ impl Service<zn::Request> for Inbound {
...
+                    // failed oneshot back so we'll fail again in case
+                    // poll_ready is called after failure.
+                    self.network_setup = Some(rx);
+                    return Poll::Ready(Err(e.into()));
                 }
-            }
+            };
+        }
+        // Now report readiness based on readiness of the inner services, if they're available.
+        // XXX do we want to propagate backpressure from the network here?
```
Comment (teor2345) [inline zebrad/src/components/inbound.rs:102]: Let's add a ticket (or a checklist item on a tracking ticket) for future work

**Review by teor2345 — DISMISSED** [review]
Looks good to me!

File: zebra-network/src/peer_set/candidate_set.rs (hunk 1 above)
Comment (dconnolly) [reply zebra-network/src/peer_set/candidate_set.rs:110]: I'm fine with keeping the loop if we add the comment 👍

**Review by dconnolly — APPROVED** [review]

File: zebrad/src/components/inbound.rs (hunk 5 above)
Comment (hdevalence) [reply zebrad/src/components/inbound.rs:102]: I'll add this to #889.

**Pull request merged**